fix(job-server): JobSQLDAO: Move blobs to a separate table (#876)

msohn opened the pull request:
* Separate the BINARIES table into BINARIES and BINARIES_CONTENTS, so
that only the BINARIES_CONTENTS table contains the "BINARY" (Blob)
column. Both tables can be joined on the same BIN_ID.

* Optimize the JobDAO.getLastUploadTimeAndType(...) method - make it
abstract and implement an optimized version in the JobFileDAO and
JobSqlDAO.

This is the result of rebasing pull request #886 onto current master and
reimplementing the migration in JDBC migration since blob columns cannot
be migrated using a SQL script.

**Pull Request checklist**

- [X] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [X] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ? not needed

**Current behavior :** 
Uploaded jars are stored in blob column "BINARY" in table "BINARIES".

**New behavior :**
Uploaded jars are stored in blob column "BINARY" in separate table "BINARIES_CONTENTS" which only contains BIN_ID (primary key) and the blob column.

**BREAKING CHANGES**
No breaking changes

**Other information**:
Existing blobs are migrated from BINARIES to BINARIES_CONTENTS

msohn commented:
I added empty line at the end of the files and added the missing try catch rollback in case something goes wrong while running migration

velvia commented on job-server/src/main/scala/db/h2/migration/V0_7_4/V0_7_4__Migrate_Blobs.scala:
```diff
@@ -0,0 +1,69 @@
+import slick.jdbc.SetParameter
+import spark.jobserver.slick.unmanaged.UnmanagedDatabase
+import java.sql.Blob
+import scala.util.control.NonFatal
```
minor nit: please group the scala imports together, in the second section (lines 5-8)

velvia commented on job-server/src/main/scala/db/postgresql/migration/V0_7_4/V0_7_4__Migrate_Blobs.scala:
```diff
@@ -0,0 +1,72 @@
+      }
+    }
+    val blob = new SerialBlob(b.binary.getBytes(1, b.binary.length().toInt))
+    val insertBlob = sqlu"""INSERT INTO "BINARIES_CONTENTS" ("BIN_ID", "BINARY") VALUES (${b.id}, ${blob})"""
```
I wish there was a way to unify these different migrations for different DBs together.  It seems most of the code is really identical, other than the SQL statements are a tiny bit different.

msohn commented:
we also need to handle deletion of BINARIES_CONTENTS when the corresponding BINARIES record is deleted

msohn commented:
* rebased onto origin/master
* fixed missing quoting in v0_7_4 migration SQL statements which crashed this migration on postgres
* reordered imports
* factored out database independent migration common code for migration v0_7_5 into trait "Migration"
* fixed deletion of jars which needs to take into account that content is now stored in table "BINARIES_CONTENTS"

msohn commented:
please consider to merge this PR soon, I'd like to avoid a third rebase onto yet another db migration

noorul commented:
@msohn Merged! Thank you!

msohn commented:
Katja will soon push another PR to avoid storage of duplicate identical blobs in the BINARIES_CONTENTS table which currently can be created if the same jar is uploaded multiple times

velvia commented:
Sounds great, thanks




> On Oct 20, 2017, at 12:29 AM, msohn <notifications@github.com> wrote:
> 
> Katja will soon push another PR to avoid storage of duplicate identical blobs in the BINARIES_CONTENTS table which currently can be created if the same jar is uploaded multiple times
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/932#issuecomment-338128158>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABA324T1ZkgYDDS_mPoniNOOTCKAfrmZks5suEvDgaJpZM4Pi9pQ>.
>
